### ISSUE TITLE
[question] cursor shape in terminal

### ISSUE THREAD

**Issue by ninrod** [syl20bnr/spacemacs#7112]
I love the terminal, I don't really know why. And emacs really does behave well in that environment, albeit with only 256 colours. 

But an annoying limitation is the cursor shape. In gui mode, cursor shape is dynamically altered to a line when user enters insert mode. It is a block shape in normal mode. It becomes a underscore in replace mode. 

Is there a way to activate that behaviour in terminal mode?

**Comment by d12frosted** [comment]
AFAIK, there is solution for this, but it doesn't work 100%.

**Comment by TheBB** [comment]
If you can fix https://github.com/7696122/evil-terminal-cursor-changer/issues/8 we can revert https://github.com/syl20bnr/spacemacs/pull/2972 which fixed Spacemacs issues https://github.com/syl20bnr/spacemacs/issues/2707, https://github.com/syl20bnr/spacemacs/issues/2940 and https://github.com/syl20bnr/spacemacs/issues/2964.

**Comment by ninrod** [comment]
Ok. I'll try to fix it when I have time available.

**Comment by kangjianbin** [comment]
My code to set cursor shape:

(defun my-set-cursor (spec)
  (if (display-graphic-p)
      (set cursor-type spec)
    (unless (equal cursor-type spec)
      (let ((shape (or (car-safe spec) spec))
            (param))
        (setq param
              (cond ((eq shape 'bar) "6")
                    ((eq shape 'hbar) "4")
                    (t "2")))
        (send-string-to-terminal
         (concat "\e[" param " q"))))))

You need a terminal that support cursor shape change. 
Looks like there is no escape sequence to support cursor width.

**Comment by ninrod** [comment]
Is that string escape sequence `(concat "\e[" param " q"))))))` standard across terminals? I think iterm2' sequence is different from mintty which is different from terminator, etc... Seems like hell.

how does emacs knows what appropriate terminal sequence escape given the specific terminal that is running emacs? 

Further, how would emacs know what specific terminal it is running under?

**Comment by TheBB** [comment]
The `TERM` environment variable is supposed to identify the terminal, if only terminals actually bothered to set it properly. Tools like terminfo and termcap can use this to identify which features the terminals have, if only systems had them consistently installed. Terminals are indeed hell, which is why I have to shake my head when people use them and *still* expect GUI-like features.

**Comment by ninrod** [comment]
So if TERM=xterm256-color, then I can be positively sure that to change the cursor I can use the same string escape sequence in ALL terminals? 

Well that's more manageable. Now I can see it working.

**Comment by kangjianbin** [comment]
Looks like set TERM environment doesn't work either. 
(getenv "TERM") always returns "dumb" in my emacs.

**Comment by elvis-sik** [comment]
Would it be possible to change the color of the cursor as well, like in the GUI?

**Comment by siulkilulki** [comment]
For vte based terminal (termite) this works for me:
  `(add-hook 'evil-insert-state-entry-hook (lambda () (send-string-to-terminal "\033[5 q")))                                                                                   
  (add-hook 'evil-normal-state-entry-hook (lambda () (send-string-to-terminal "\033[0 q")))                                                                                   `
0, 2- box
1 - blinking box  
3 - blinking underline
4 - steady underline
5 - blinking bar
6 - steady bar
Try it by yourself in terminal with e.g. `printf '\033[5 q'`
also see https://vt100.net/docs/vt510-rm/DECSCUSR

**Comment by SlightlyCyborg** [comment]
@siulkilulki is the MVP

**Comment by h0d** [comment]
I just created a package in order to address this issue. It does not depend on `evil-mode` but currently only supports VT520-based terminals. Feedback and contribution are welcome.

see [term-cursor.el](/h0d/term-cursor.el)

**Comment by arronmabrey** [comment]
@h0d Thanks your package is working for me in iTerm 3.x 🤗

```
dotspacemacs-additional-packages '((term-cursor :location (recipe :fetcher github :repo "h0d/term-cursor.el" )))
```

**Comment by cinerea0** [comment]
This issue probably shouldn't be closed, since the fix has not been integrated into spacemacs.

**Comment by mskar** [comment]
Hi everyone,
I'm a long time vim user. I just tried spacemacs today and I am blown away.
Everything works well in the GUI and terminal without additional setup except for the cursor shape in terminal emacs.
Luckily, I found this issue, installed @h0d's package using the line provided by @arronmabrey and then added `(global-term-cursor-mode)` to my user config as below.
```lisp
(defun dotspacemacs/user-config ()
  "Configuration for user code:
This function is called at the very end of Spacemacs startup, after layer
configuration.
Put your configuration code here, except for variables that should be set
before packages are loaded."
  (global-term-cursor-mode)
  )
```
Thanks @h0d and @arronmabrey for the fix.
Thanks to everyone for making spacemacs great!
